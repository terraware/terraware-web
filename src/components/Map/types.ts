import { Feature, FeatureCollection, GeoJsonProperties, MultiPolygon, Polygon } from 'geojson';

export type ReadOnlyBoundary = {
  featureCollection: FeatureCollection;
  id: string;
};

export type GeometryFeature = Feature<Polygon | MultiPolygon, GeoJsonProperties>;
