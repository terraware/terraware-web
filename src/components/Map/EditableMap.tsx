import React, { useCallback, useEffect, useRef, useState } from 'react';
import ReactMapGL, { FullscreenControl, LngLatBoundsLike, MapRef } from 'react-map-gl';

import { Box, Typography, useTheme } from '@mui/material';
import bbox from '@turf/bbox';
import { MultiPolygon } from 'geojson';

import EditableMapDraw, { MapEditorMode } from 'src/components/Map/EditableMapDraw';
import { useIsVisible } from 'src/hooks/useIsVisible';
import strings from 'src/strings';
import { MapViewStyles } from 'src/types/Map';
import useMapboxToken from 'src/utils/useMapboxToken';

import MapViewStyleControl, { useMapViewStyle } from './MapViewStyleControl';

export type EditableMapProps = {
  boundary?: MultiPolygon;
  onBoundaryChanged: (boundary: MultiPolygon | null) => void;
  style?: object;
};

const polygonIcon = <button className='mapbox-gl-draw_polygon mapbox-gl-draw_ctrl-draw-btn' />;
const trashIcon = <button className='mapbox-gl-draw_trash mapbox-gl-draw_ctrl-draw-btn' />;

function instructionsForMode(mode: MapEditorMode | null): JSX.Element[] {
  let instructions: string = '';

  if (mode === 'ReplacingBoundary') {
    instructions = strings.MAP_EDITOR_DRAWING_ADDITIONAL_FEATURE;
  } else if (mode === 'CreatingBoundary') {
    instructions = strings.MAP_EDITOR_DRAWING_FIRST_FEATURE;
  } else if (mode === 'EditingBoundary') {
    instructions = strings.MAP_EDITOR_EDITING_FEATURE;
  } else if (mode === 'NoBoundary') {
    instructions = strings.MAP_EDITOR_NO_FEATURES;
  } else if (mode === 'BoundaryNotSelected') {
    instructions = strings.MAP_EDITOR_NO_SELECTION;
  } else if (mode === 'BoundarySelected') {
    instructions = strings.MAP_EDITOR_SELECTED_FEATURE;
  } else if (mode === null) {
    instructions = strings.LOADING;
  }

  return strings.formatString(instructions, {
    polygon: polygonIcon,
    trash: trashIcon,
  }) as JSX.Element[];
}

export default function EditableMap({ boundary, onBoundaryChanged, style }: EditableMapProps): JSX.Element {
  const { mapId, refreshToken, token } = useMapboxToken();
  const [firstVisible, setFirstVisible] = useState<boolean>(false);
  const [mode, setMode] = useState<MapEditorMode | null>(null);
  const containerRef = useRef(null);
  const mapRef = useRef<MapRef | null>(null);
  const visible = useIsVisible(containerRef);
  const theme = useTheme();
  const [mapViewStyle, onChangeMapViewStyle] = useMapViewStyle();

  useEffect(() => {
    // `firstVisible` detects when the box containing the map is first visible in the viewport. The map should only be
    // rendered if `firstVisible` is true. This accounts for cases in which the map is initially rendered hidden, and
    // is improperly resized when it first becomes visible.
    setFirstVisible((fv) => fv || visible);
  }, [visible]);

  const onMapError = useCallback(
    (event: any) => {
      if (event?.error?.status === 401 && refreshToken) {
        refreshToken();
      }
    },
    [refreshToken]
  );

  const initialViewState = {
    bounds: boundary ? (bbox(boundary) as LngLatBoundsLike) : undefined,
    fitBoundsOptions: {
      animate: false,
      padding: 25,
    },
  };

  return (
    <Box ref={containerRef} display='flex' flexDirection='column' flexGrow={1}>
      {firstVisible && (
        <>
          <Typography fontSize='16px' margin={theme.spacing(1, 0)} minHeight='4em'>
            {instructionsForMode(mode)}
          </Typography>
          <ReactMapGL
            key={mapId}
            onError={onMapError}
            ref={mapRef}
            mapboxAccessToken={token}
            mapStyle={MapViewStyles[mapViewStyle]}
            style={{
              display: 'flex',
              flexGrow: '1',
              flexDirection: 'column',
              height: 'auto',
              minHeight: '436px',
              ...style,
            }}
            initialViewState={initialViewState}
          >
            <FullscreenControl position='top-left' />
            <MapViewStyleControl mapViewStyle={mapViewStyle} onChangeMapViewStyle={onChangeMapViewStyle} />
            <EditableMapDraw boundary={boundary} onBoundaryChanged={onBoundaryChanged} setMode={setMode} />
          </ReactMapGL>
        </>
      )}
    </Box>
  );
}
