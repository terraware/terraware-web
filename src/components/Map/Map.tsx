import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { Box } from '@mui/material';
import 'mapbox-gl/dist/mapbox-gl.css';
import ReactMapGL, { AttributionControl, Layer, NavigationControl, Popup, Source } from 'react-map-gl';
import { MapSource, MapEntityId, MapEntityOptions, MapOptions, MapPopupRenderer } from 'src/types/Map';
import { MapService } from 'src/services';

/**
 * The following is needed to deal with a mapbox bug
 * See: https://docs.mapbox.com/mapbox-gl-js/guides/install/#transpiling
 */
import mapboxgl from 'mapbox-gl';
import MapBanner from './MapBanner';
const mapboxImpl: any = mapboxgl;
// @tslint
// eslint-disable-next-line import/no-webpack-loader-syntax
mapboxImpl.workerClass = require('worker-loader!mapbox-gl/dist/mapbox-gl-csp-worker').default; /* tslint:disable-line */

const navControlStyle = {
  marginRight: '5px',
  marginBottom: '20px',
};

type PopupInfo = {
  lng: number;
  lat: number;
  properties: any;
  sourceId: string;
};

export type MapProps = {
  token: string;
  options: MapOptions;
  onTokenExpired?: () => void;
  popupRenderer?: MapPopupRenderer;
  // changes to mapId will re-create the map, needed for new token refreshes
  // since mapbox token is not reactive
  mapId?: string;
  // style overrides
  style?: object;
  bannerMessage?: string;
  // entity options
  entityOptions?: MapEntityOptions;
};

export default function Map(props: MapProps): JSX.Element {
  const { token, onTokenExpired, options, popupRenderer, mapId, style, bannerMessage, entityOptions } = props;
  const [geoData, setGeoData] = useState<any[]>();
  const [layerIds, setLayerIds] = useState<string[]>([]);
  const [popupInfo, setPopupInfo] = useState<PopupInfo | null>(null);
  const [deferredHighlightEntity, setDeferredHighlightEntity] = useState<MapEntityId | undefined>();
  const [deferredFocusEntity, setDeferredFocusEntity] = useState<MapEntityId | undefined>();
  const mapRef = useRef(null);
  const hoverStateId: { [key: string]: number | undefined } = useMemo(() => ({}), []);
  const selectStateId: { [key: string]: number | undefined } = useMemo(() => ({}), []);
  const highlightStateId: { [key: string]: number | undefined } = useMemo(() => ({}), []);

  const getFillColor = (source: MapSource, type: 'highlight' | 'select' | 'hover' | 'default') => {
    switch (type) {
      case 'highlight':
        return source.highlightFillColor || source.fillColor;
      case 'select':
        return source.selectFillColor || source.fillColor;
      case 'hover':
        return source.hoverFillColor || source.fillColor;
      default:
        return source.fillColor;
    }
  };

  const onMapError = useCallback(
    (event: any) => {
      if (event?.error?.status === 401) {
        // tslint:disable-next-line: no-console
        console.error('Mapbox token expired');
        if (onTokenExpired) {
          onTokenExpired();
        }
      }
    },
    [onTokenExpired]
  );

  const updateFeatureState = useCallback((featureVar: any, property: string, mapEntityId: MapEntityId) => {
    const map: any = mapRef && mapRef.current;
    if (!map) {
      return;
    }
    const { id, sourceId } = mapEntityId;
    // clear previous
    if (featureVar[sourceId] && id !== featureVar[sourceId]) {
      map.setFeatureState({ source: sourceId, id: featureVar[sourceId] }, { [property]: false });
    }
    // set new
    if (id) {
      map.setFeatureState({ source: sourceId, id }, { [property]: true });
    }
    featureVar[sourceId] = id;
  }, []);

  const onMapClick = useCallback(
    (event: any) => {
      const { lat, lng } = event.lngLat;
      if (event?.features[0]?.properties) {
        const { id, properties, layer } = event.features[0];
        const sourceId = layer.id.replace(/-fill$/, '');
        setPopupInfo({
          lng,
          lat,
          properties,
          sourceId,
        });
        updateFeatureState(selectStateId, 'select', { sourceId, id });
      }
    },
    [updateFeatureState, selectStateId]
  );

  const drawFocusTo = useCallback(
    (mapEntity: MapEntityId) => {
      const map: any = mapRef?.current;
      if (!map || !mapEntity.id || !mapEntity.sourceId) {
        return;
      }
      const foundSource = geoData && geoData.find((geo) => geo.id === mapEntity.sourceId);
      if (!foundSource || !foundSource.data) {
        return;
      }
      const feature = foundSource.data.features.find((featureData: any) => featureData.id === mapEntity.id);
      if (!feature) {
        return;
      }
      const coordinates = feature.geometry.coordinates;
      const bbox = MapService.getBoundingBox([[coordinates]]);
      map.fitBounds([bbox.lowerLeft, bbox.upperRight], { padding: 20 });
    },
    [geoData]
  );

  const onLoad = () => {
    const map: any = mapRef && mapRef.current;
    if (!map) {
      return;
    }
    Object.keys(hoverStateId).forEach((key) => delete hoverStateId[key]);
    Object.keys(selectStateId).forEach((key) => delete selectStateId[key]);
    Object.keys(highlightStateId).forEach((key) => delete highlightStateId[key]);

    const { sources } = options;
    sources
      .filter((source) => source.isInteractive)
      .forEach((source) => {
        const layerId = `${source.id}-fill`;
        map.on('mousemove', layerId, (e: any) => {
          if (!e.features.length || !e.features[0].properties) {
            return;
          }
          const newId = e.features[0].id;
          updateFeatureState(hoverStateId, 'hover', { sourceId: source.id, id: newId });
        });
        map.on('mouseleave', `${source.id}-fill`, () => {
          updateFeatureState(hoverStateId, 'hover', { sourceId: source.id });
        });
      });

    if (deferredHighlightEntity) {
      updateFeatureState(highlightStateId, 'highlight', deferredHighlightEntity);
      setDeferredHighlightEntity(undefined);
    }
    if (deferredFocusEntity) {
      drawFocusTo(deferredFocusEntity);
      setDeferredFocusEntity(undefined);
    }
  };

  useEffect(() => {
    const { sources } = options;
    if (geoData) {
      return;
    }

    /**
     * Initialize sources - we want ONE source per data type, eg. one source containing all geometries for zones.
     * Creating multiple sources, one per zone or one per subzone, is unnecessarily and will create performance
     * bottlenecks. Leverage data-drive rendering using properties in the features and mapbox supported properties:
     *  eg. 'text-field': ['get', source.annotation.textField]
     */
    const geo = sources
      .map((source) => {
        const features = source.entities
          .map((entity) => {
            const multiPolygons = MapService.getMapEntityGeometry(entity);
            if (!multiPolygons.length) {
              return null;
            }
            return multiPolygons.map((multiPolygon) => ({
              type: 'Feature',
              geometry: {
                type: 'Polygon',
                coordinates: multiPolygon,
              },
              properties: entity.properties,
              id: entity.id,
            }));
          })
          .filter((f) => f)
          .flatMap((f) => f);

        return {
          id: source.id,
          isInteractive: source.isInteractive,
          data: {
            type: 'FeatureCollection',
            features,
          },
          layer: {
            id: `${source.id}-fill`,
            type: 'fill',
            paint: {
              'fill-color': [
                // Use a case-expression to decide which color to use for fill
                // see https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#case
                // and, https://docs.mapbox.com/mapbox-gl-js/api/map/#map#setfeaturestate
                // if highlight, use highlight color
                // else if selected, use select color
                // else if hover, user hover color
                // else use default fill color
                'case',
                ['boolean', ['feature-state', 'highlight'], false],
                getFillColor(source, 'highlight'),
                ['boolean', ['feature-state', 'select'], false],
                getFillColor(source, 'select'),
                ['boolean', ['feature-state', 'hover'], false],
                getFillColor(source, 'hover'),
                getFillColor(source, 'default'),
              ],
            },
          },
          layerOutline: {
            id: `${source.id}-outline`,
            type: 'line',
            paint: {
              'line-color': source.lineColor,
              'line-width': source.lineWidth,
            },
          },
          textAnnotation: source.annotation
            ? {
                id: `${source.id}-annotation`,
                type: 'symbol',
                paint: {
                  'text-color': source.annotation.textColor,
                },
                layout: {
                  'text-allow-overlap': false,
                  'text-ignore-placement': false,
                  'text-field': ['get', source.annotation.textField],
                  'text-anchor': 'center',
                  'text-size': source.annotation.textSize,
                },
              }
            : null,
        };
      })
      .filter((g) => g);
    setGeoData(geo as any);
    if (popupRenderer) {
      setLayerIds(geo.filter((g: any) => g.isInteractive).map((g: any) => g.layer.id));
    }
  }, [options, geoData, setGeoData, token, popupRenderer]);

  useEffect(() => {
    if (entityOptions?.highlight) {
      if (!mapRef || !mapRef.current) {
        setDeferredHighlightEntity(entityOptions?.highlight);
      } else {
        updateFeatureState(highlightStateId, 'highlight', entityOptions?.highlight);
      }
    }
  }, [entityOptions?.highlight, highlightStateId, updateFeatureState]);

  useEffect(() => {
    if (entityOptions?.focus) {
      if (!mapRef || !mapRef.current) {
        setDeferredFocusEntity(entityOptions?.focus);
      } else {
        drawFocusTo(entityOptions?.focus);
      }
    }
  }, [drawFocusTo, entityOptions?.focus]);

  const mapSources = useMemo(() => {
    if (!geoData) {
      return null;
    }
    const sources = (geoData as any[]).map((geo: any, index) => (
      <Source type='geojson' key={index} data={geo.data} id={geo.id}>
        {geo.textAnnotation && <Layer {...geo.textAnnotation} />}
        {geo.layerOutline && <Layer {...geo.layerOutline} />}
        {geo.layer && <Layer {...geo.layer} />}
      </Source>
    ));

    return sources;
  }, [geoData]);

  const hasEntities = options.sources?.some((source) => {
    return source.entities?.some((entity) => entity?.boundary?.length);
  });

  return (
    <Box sx={{ display: 'flex', flexGrow: 1, height: '100%', minHeight: 250, position: 'relative' }}>
      {bannerMessage && <MapBanner message={bannerMessage} />}
      <ReactMapGL
        key={mapId}
        mapboxAccessToken={token}
        mapStyle='mapbox://styles/mapbox/satellite-v9?optimize=true'
        initialViewState={{
          bounds: hasEntities ? [options.bbox.lowerLeft, options.bbox.upperRight] : undefined,
          fitBoundsOptions: hasEntities ? { padding: 20 } : undefined,
        }}
        interactiveLayerIds={layerIds}
        onError={onMapError}
        onClick={onMapClick}
        style={style}
        attributionControl={false}
        onLoad={onLoad}
        ref={mapRef}
        onRender={(event) => event.target.resize()}
      >
        {mapSources}
        <NavigationControl showCompass={false} style={navControlStyle} position='bottom-right' />
        <AttributionControl compact={true} style={{ marginRight: '5px' }} />
        {popupInfo && popupRenderer && (
          <Popup
            anchor='top'
            longitude={Number(popupInfo.lng)}
            latitude={Number(popupInfo.lat)}
            onClose={() => {
              setPopupInfo(null);
              updateFeatureState(selectStateId, 'select', { sourceId: popupInfo.sourceId });
            }}
            style={popupRenderer.style}
            className={popupRenderer.className}
          >
            {popupRenderer.render(popupInfo.properties)}
          </Popup>
        )}
      </ReactMapGL>
    </Box>
  );
}
