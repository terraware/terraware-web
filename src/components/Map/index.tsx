import GenericMap from './GenericMap';
import Map from './Map';
import { useSpeciesPlantsRenderer } from './MapRenderUtils';
import PlantingSiteMap from './PlantingSiteMap';

export { Map, PlantingSiteMap, GenericMap, useSpeciesPlantsRenderer };
