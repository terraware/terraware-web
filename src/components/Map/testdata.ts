import { FeatureCollection, Geometry } from 'geojson';
import { GeometryFeature } from 'src/types/Map';

export const feature1: GeometryFeature = {
  type: 'Feature',
  geometry: {
    type: 'MultiPolygon',
    coordinates: [
      [
        [
          [5, 5],
          [10, 5],
          [10, 10],
          [5, 10],
          [5, 5],
        ],
      ],
    ],
  },
  properties: { hello: 'world' },
  id: 0,
};

export const feature2: GeometryFeature = {
  type: 'Feature',
  geometry: {
    type: 'MultiPolygon',
    coordinates: [
      [
        [
          [10, 5],
          [15, 5],
          [15, 10],
          [10, 10],
          [10, 5],
        ],
      ],
    ],
  },
  properties: { yoyo: 'ma' },
  id: 1,
};

export const feature3: GeometryFeature = {
  type: 'Feature',
  geometry: {
    type: 'MultiPolygon',
    coordinates: [
      [
        [
          [15, 5],
          [20, 5],
          [20, 10],
          [15, 10],
          [15, 5],
        ],
      ],
    ],
  },
  properties: { lorem: 'ipsum' },
  id: 2,
};

export const cutOnNoOverlap: Geometry = {
  type: 'MultiPolygon',
  coordinates: [
    [
      [
        [100, 5],
        [150, 5],
        [150, 10],
        [100, 10],
        [100, 5],
      ],
    ],
  ],
};

export const cutWithOverlap: Geometry = {
  type: 'MultiPolygon',
  coordinates: [
    [
      [
        [7.5, 0],
        [12.5, 0],
        [12.5, 100],
        [7.5, 100],
        [7.5, 0],
      ],
    ],
  ],
};

export const featureCollection1: FeatureCollection = { type: 'FeatureCollection', features: [feature1] };
export const featureCollection2: FeatureCollection = { type: 'FeatureCollection', features: [feature1, feature2] };

export const cutOnNoOverlapFeature: GeometryFeature = {
  type: 'Feature',
  geometry: cutOnNoOverlap,
  properties: {},
  id: 0,
};
export const cutOnOverlapFeature: GeometryFeature = {
  type: 'Feature',
  geometry: cutWithOverlap,
  properties: {},
  id: 1,
};
