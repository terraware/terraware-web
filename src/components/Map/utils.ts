import { Feature, FeatureCollection, Geometry, MultiPolygon } from 'geojson';
import intersect from '@turf/intersect';
import difference from '@turf/difference';
import _ from 'lodash';
import { MapSourceProperties, MapSourceRenderProperties } from 'src/types/Map';
import { GeometryFeature } from './types';

export function toMultiPolygon(geometry: Geometry): MultiPolygon | null {
  if (geometry.type === 'MultiPolygon') {
    return geometry;
  } else if (geometry.type === 'Polygon') {
    return { type: 'MultiPolygon', coordinates: [geometry.coordinates] };
  } else {
    return null;
  }
}

export function toMultiPolygonArray(featureCollection: FeatureCollection): MultiPolygon[] | undefined {
  const polyArray = featureCollection.features
    .map((feature: Feature) => toMultiPolygon(feature.geometry))
    .filter((poly: MultiPolygon | null) => poly !== null) as MultiPolygon[];

  return polyArray.length ? polyArray : undefined;
}

export function toFeature(
  multiPolygon: MultiPolygon,
  properties: MapSourceProperties,
  id: number
): Feature & { id: number } {
  return {
    type: 'Feature',
    geometry: {
      type: 'MultiPolygon',
      coordinates: multiPolygon.coordinates,
    },
    properties,
    id,
  };
}

export const getFillColor = (source: MapSourceRenderProperties, type: 'highlight' | 'select' | 'hover' | 'default') => {
  switch (type) {
    case 'highlight':
      return source.highlightFillColor || source.fillColor;
    case 'select':
      return source.selectFillColor || source.fillColor;
    case 'hover':
      return source.hoverFillColor || source.fillColor;
    default:
      return source.fillColor;
  }
};

export const getMapDrawingLayer = (source: MapSourceRenderProperties, sourceId: string) => {
  return {
    id: sourceId,
    isInteractive: source.isInteractive,
    layer: {
      id: `${sourceId}-fill`,
      type: 'fill',
      paint: {
        'fill-color': [
          // Use a case-expression to decide which color to use for fill
          // see https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#case
          // and, https://docs.mapbox.com/mapbox-gl-js/api/map/#map#setfeaturestate
          // if highlight, use highlight color
          // else if selected, use select color
          // else if hover, user hover color
          // else use default fill color
          'case',
          ['boolean', ['feature-state', 'highlight'], false],
          getFillColor(source, 'highlight'),
          ['boolean', ['feature-state', 'select'], false],
          getFillColor(source, 'select'),
          ['boolean', ['feature-state', 'hover'], false],
          getFillColor(source, 'hover'),
          getFillColor(source, 'default'),
        ],
      },
    },
    layerOutline: {
      id: `${sourceId}-outline`,
      type: 'line',
      paint: {
        'line-color': source.lineColor,
        'line-width': source.lineWidth,
      },
    },
    textAnnotation: source.annotation
      ? {
          id: `${sourceId}-annotation`,
          type: 'symbol',
          paint: {
            'text-color': source.annotation.textColor,
          },
          layout: {
            'text-allow-overlap': false,
            'text-ignore-placement': false,
            'text-field': ['get', source.annotation.textField],
            'text-anchor': 'center',
            'text-size': source.annotation.textSize,
          },
        }
      : null,
    patternFill: source.patternFill
      ? {
          id: `${sourceId}-pattern`,
          type: 'fill',
          paint: {
            'fill-pattern': source.patternFill.imageName,
            'fill-opacity': source.patternFill.opacityExpression ?? 1.0,
          },
        }
      : null,
  };
};

/**
 * Cut source polygons into smaller polygons
 * @param source
 *  The source polygons that need to be split into smaller polygons
 * @param cutWith
 *  The polygon to use as the overlap check in order to cut polygons
 * @returns
 *  List of polygons that were cut as a result of overlaps with cutWith
 *  Null if there was no overlap
 */
export const cutPolygons = (source: GeometryFeature[], cutWith: Geometry): GeometryFeature[] | null => {
  if (cutWith.type !== 'Polygon' && cutWith.type !== 'MultiPolygon') {
    return null;
  }

  const intersections = source.map((poly) => intersect(cutWith, poly));
  if (!intersections.some((f) => f !== null)) {
    return null;
  }

  const splitFeatures = intersections.reduce((acc, curr, index) => {
    const originalFeature = source[index];
    if (curr !== null) {
      const subtracted = originalFeature.properties?.isFixed ? null : difference(originalFeature, curr);
      if (subtracted !== null) {
        return [...acc, subtracted, curr];
      }
    }
    return [...acc, originalFeature];
  }, [] as GeometryFeature[]);

  return _.isEqual(splitFeatures, source) ? null : splitFeatures;
};
