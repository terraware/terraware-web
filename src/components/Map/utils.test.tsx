/**
 * @jest-environment node
 */
import { Feature } from 'geojson';

import { ReadOnlyBoundary, RenderableReadOnlyBoundary } from 'src/types/Map';

import { cutOnNoOverlap, cutWithOverlap, feature1, feature2, feature3 } from './testdata';
import { boundariesToViewState, leftMostFeature, overlayAndSubtract, readOnlyBoundariesToMapLayers } from './utils';

describe('Map utils', () => {
  describe('leftMostFeature', () => {
    test("should return null when there isn't a left most feature", () => {
      expect(leftMostFeature([])).toBeNull();
    });

    test("should return a left most feature with it's center point", () => {
      const leftMost = leftMostFeature([feature1, feature2]);

      expect(leftMost?.feature?.id).toBe(0);
      expect(leftMost?.center?.[0]).toBe(7.5);
      expect(leftMost?.center?.[1]).toBe(7.5);
    });
  });

  describe('overlayAndSubtract', () => {
    test('should return null when the overlay geometry is not a polygon', () => {
      expect(overlayAndSubtract([feature1, feature2], { type: 'Point', coordinates: [1, 2] })).toBeNull();
    });

    test('should return null when the overlay geometry does not overlap with the input', () => {
      expect(overlayAndSubtract([feature1, feature2], cutOnNoOverlap)).toBeNull();
    });

    test('should return existing polygons without the intersection, and the new polygon consisting of + the intersection, when there is an overlap with cutting geometry', () => {
      const expected = [
        {
          type: 'Feature',
          geometry: {
            coordinates: [
              [
                [
                  [5, 5],
                  [7.5, 5],
                  [7.5, 10],
                  [5, 10],
                  [5, 5],
                ],
              ],
            ],
            type: 'MultiPolygon',
          },
          id: 0,
          properties: { hello: 'world' },
        },
        {
          type: 'Feature',
          geometry: {
            coordinates: [
              [
                [
                  [12.5, 5],
                  [15, 5],
                  [15, 10],
                  [12.5, 10],
                  [12.5, 5],
                ],
              ],
            ],
            type: 'MultiPolygon',
          },
          id: 1,
          properties: { yoyo: 'ma' },
        },
        {
          type: 'Feature',
          geometry: {
            coordinates: [
              [
                [
                  [15, 5],
                  [20, 5],
                  [20, 10],
                  [15, 10],
                  [15, 5],
                ],
              ],
            ],
            type: 'MultiPolygon',
          },
          id: 2,
          properties: { lorem: 'ipsum' },
        },
        {
          type: 'Feature',
          geometry: {
            coordinates: [
              [
                [
                  [7.5, 5],
                  [12.5, 5],
                  [12.5, 10],
                  [7.5, 10],
                  [7.5, 5],
                ],
              ],
            ],
            type: 'MultiPolygon',
          },
          properties: {},
        },
      ];

      expect(overlayAndSubtract([feature1, feature2, feature3], cutWithOverlap)).toStrictEqual(expected);
    });
  });

  describe('boundariesToViewState', () => {
    test('should view the earth when there are no boundaries', () => {
      expect(boundariesToViewState([])).toEqual({
        bounds: [Infinity, Infinity, -Infinity, -Infinity],
        fitBoundsOptions: {
          animate: false,
          padding: 25,
        },
      });
    });

    test('has the correct bounds', () => {
      const boundaries: ReadOnlyBoundary[] = [
        {
          data: {
            type: 'FeatureCollection',
            features: [
              {
                type: 'Feature',
                geometry: {
                  type: 'MultiPolygon',
                  coordinates: [
                    [
                      [
                        [32.81043714, 2.24535757],
                        [33.02329606, 2.64270388],
                        [33.22495188, 2.5084025],
                        [33.05130381, 2.29573331],
                        [32.81043714, 2.24535757],
                      ],
                    ],
                  ],
                },
                properties: { id: 'applicationBoundary' },
                id: 'applicationBoundary',
              },
            ],
          },
          id: 'applicationBoundary',
        },
      ];
      expect(boundariesToViewState(boundaries)).toEqual({
        bounds: [32.81043714, 2.24535757, 33.22495188, 2.64270388],
        fitBoundsOptions: {
          animate: false,
          padding: 25,
        },
      });
    });
  });

  describe('readOnlyBoundariesToMapLayers', () => {
    test("should return null when there aren't any boundaries", () => {
      expect(readOnlyBoundariesToMapLayers([])).toBeNull();
      expect(readOnlyBoundariesToMapLayers(undefined)).toBeNull();
    });

    test('generates map layers from boundaries', () => {
      const boundary: Feature = {
        geometry: {
          type: 'MultiPolygon',
          coordinates: [
            [
              [
                [32.81043714, 2.24535757],
                [33.02329606, 2.64270388],
                [33.22495188, 2.5084025],
                [33.05130381, 2.29573331],
                [32.81043714, 2.24535757],
              ],
            ],
          ],
        },
        id: 'countryBoundary',
        properties: { id: 'countryBoundary' },
        type: 'Feature',
      };
      const boundaries: RenderableReadOnlyBoundary[] = [
        {
          id: '1',
          data: {
            type: 'FeatureCollection',
            features: [boundary],
          },
          renderProperties: {
            isInteractive: false,
            fillColor: 'highlight',
            lineColor: 'blue',
            lineWidth: 1,
          },
        },
      ];

      expect(JSON.stringify(readOnlyBoundariesToMapLayers(boundaries))).toMatchSnapshot();
    });
  });
});
