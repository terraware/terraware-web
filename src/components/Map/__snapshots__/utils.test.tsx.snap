// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map utils readOnlyBoundariesToMapLayers generates map layers from boundaries 1`] = `"[{\\"key\\":\\"1\\",\\"ref\\":null,\\"props\\":{\\"type\\":\\"geojson\\",\\"data\\":{\\"type\\":\\"FeatureCollection\\",\\"features\\":[{\\"geometry\\":{\\"type\\":\\"MultiPolygon\\",\\"coordinates\\":[[[[32.81043714,2.24535757],[33.02329606,2.64270388],[33.22495188,2.5084025],[33.05130381,2.29573331],[32.81043714,2.24535757]]]]},\\"id\\":\\"countryBoundary\\",\\"properties\\":{\\"id\\":\\"countryBoundary\\"},\\"type\\":\\"Feature\\"}]},\\"id\\":\\"1\\",\\"children\\":[null,null,{\\"key\\":null,\\"ref\\":null,\\"props\\":{\\"id\\":\\"1-outline\\",\\"type\\":\\"line\\",\\"paint\\":{\\"line-color\\":[\\"case\\",[\\"boolean\\",[\\"feature-state\\",\\"select\\"],false],\\"blue\\",\\"blue\\"],\\"line-width\\":[\\"case\\",[\\"boolean\\",[\\"feature-state\\",\\"select\\"],false],1,1]}},\\"_owner\\":null,\\"_store\\":{}},{\\"key\\":null,\\"ref\\":null,\\"props\\":{\\"id\\":\\"1-fill\\",\\"type\\":\\"fill\\",\\"paint\\":{\\"fill-color\\":[\\"case\\",[\\"boolean\\",[\\"feature-state\\",\\"highlight\\"],false],\\"highlight\\",[\\"boolean\\",[\\"feature-state\\",\\"select\\"],false],\\"highlight\\",[\\"boolean\\",[\\"feature-state\\",\\"hover\\"],false],\\"highlight\\",\\"highlight\\"]}},\\"_owner\\":null,\\"_store\\":{}}]},\\"_owner\\":null,\\"_store\\":{}}]"`;
